src/ast/passes/codegen_llvm.cpp

```diff
@@ -2502,6 +2550,24 @@ void CodegenLLVM::createRet(Value *value)
     b_.CreateRet(value);
     return;
   }
+  else if (inside_subprog_)
+  {
+    // Subprogs have no default return value, return void for void subprogs
+    // and an annotated dummy value for other subprogs
+    // Note: programs with reachable blocks returning the dummy value are
+    // rejected later; dummy value is needed for the LLVM IR to be correct
+    auto ret_type = b_.GetInsertBlock()->getParent()->getReturnType();
+    if (ret_type->isVoidTy())
+      b_.CreateRetVoid();
+    else
+    {
+      auto dummy_alloca = b_.CreateAlloca(ret_type, nullptr, "ret_dummy");
+      auto dummy = b_.CreateLoad(ret_type, dummy_alloca);
+      auto ret_inst = b_.CreateRet(dummy);
+      ret_inst->setMetadata("ret_dummy", MDTuple::get(b_.getContext(), {}));
```

Comment:
Does moving return checks into semantic analyser get rid of the ret_dummy stuff? Not immediately clear to me what this does